apps/prairielearn/src/pages/authCallbackLti/authCallbackLti.js

```diff
@@ -15,202 +13,168 @@ const { config } = require('../../lib/config');
+    const timeDiff = Math.abs(Math.floor(Date.now() / 1000) - parameters.oauth_timestamp);
+    if (timeDiff > TIME_TOLERANCE_SEC) {
+      throw error.make(403, 'Invalid timestamp');
+    }
+
+    // Check nonce hasn't been used by that consumer_key in that timeframe
+    // https://oauth.net/core/1.0/#nonce
+    const nonceKey = 'authCallbackLti:' + parameters.oauth_timestamp + ':' + parameters.oauth_nonce;
+    const nonceVal = await cache.get(nonceKey);
+
+    if (nonceVal) {
+      throw error.make(403, 'Nonce reused');
+    }
+
+    // Remember that this nonce was already used.
+    cache.set(nonceKey, true, TIME_TOLERANCE_SEC * 1000);
+
+    // OAuth validation succeeded, next look up and store user authn data
+    if (!parameters.user_id) {
+      throw error.make(400, 'Authentication problem: UserID required. Anonymous access disabled.');
+    }
+
+    // Create unique UID from LTI parameters and CI id
+    // Not using an email address (parameters.lis_person_contact_email_primary)
+    // so that LTI doesn't conflict with other UIDs.
+    const authUid =
+      parameters.user_id + '@' + parameters.context_id + '::ciid=' + ltiresult.course_instance_id;
+
+    let fallbackName = 'LTI user';
+    if (parameters.context_title) {
+      fallbackName = `${parameters.context_title} user`;
+      // e.g. UIUC Degree Sandbox user
+    }
+    const authName = parameters.lis_person_name_full || fallbackName;
+
+    const userResult = await sqldb.callAsync('users_select_or_insert_and_enroll_lti', [
+      authUid,
+      authName,
+      ltiresult.course_instance_id,
+      parameters.user_id,
+      parameters.context_id,
+      res.locals.req_date,
+    ]);
+    if (!userResult.rows[0].has_access) {
+      throw error.make(403, 'Access denied');
+    }
+
+    const tokenData = {
+      user_id: result.rows[0].user_id,
+      authn_provider_name: 'LTI',
+    };
+    const pl_authn = generateSignedToken(tokenData, config.secretKey);
+    res.cookie('pl_authn', pl_authn, {
+      maxAge: config.authnCookieMaxAgeMilliseconds,
+      httpOnly: true,
+      secure: shouldSecureCookie(req),
+    });
+
+    // Dual-write information to the session so that we can start reading
+    // it instead of the cookie in the future.
+    req.session.user_id = userResult.rows[0].user_id;
+    req.session.authn_provider_name = 'LTI';
+
+    const linkResult = await sqldb.queryOneRowAsync(sql.upsert_current_link, {
+      course_instance_id: ltiresult.course_instance_id,
+      context_id: parameters.context_id,
+      resource_link_id: parameters.resource_link_id,
+      resource_link_title: parameters.resource_link_title || '',
+      resource_link_description: parameters.resource_link_description || '',
+    });
+
+    // Do we have an assessment linked to this resource_link_id?
+    if (linkResult.rows[0].assessment_id) {
+      if ('lis_result_sourcedid' in parameters) {
+        // Save outcomes here
+        await sqldb.queryAsync(sql.upsert_outcome, {
```

Comment:
The previous behaviour had this done in the background. Just double-checking that you're ok with awaiting for this update before redirecting to the assessment.

Reply:
This is just a database query, there's nothing that should take any noticeable amount of time here, so I'm very comfortable doing this inline.